### ISSUE TITLE
Suggestion: Add support for WG's send-chat region flag.

### ISSUE THREAD

**Issue by delbertina** [TownyAdvanced/Towny#3104]
This ticket template is for bug/error reports. If you are suggesting a feature and not reporting a bug please delete all the text here and begin your ticket Title with Suggestion:

### What steps will reproduce the problem?
1. Spigot or Paper server w/ WG, WE, 3 Towny jars, and DiscordSRV.
2. Create WG region with send-chat flag set to deny
3. Chat in region

### What is the expected output?
1. Event to be canceled and not sent to DiscordSRV chat channel
2. Persists even with modify chat disabled

### What is your Towny version?
Do not say "latest".
0.93.0.0

### What is your spigot/craftbukkit version?
Do not say "latest".
Tested on paperclip 417 and 427. Also, most current build of Spigot 1.13.2. Commit 2474d93d1f8 I think?

### Please use Pastebin.com to link the following files:

**Skipping these files is not an option. It is the easiest way to diagnose an issue ticket properly. If you do skip this step then your ticket will most likely be deleted and you will be asked to resubmit.**
1. Your full server startup from the logs\latest.log : [https://pastebin.com/h9KhX0EN](https://pastebin.com/h9KhX0EN)
2. Your Towny config.yml (if using MySQL please remove password,) : [https://pastebin.com/ZvRHMPFB](https://pastebin.com/ZvRHMPFB)
3. Your townyperms.yml : [https://pastebin.com/28Td5JH4](https://pastebin.com/28Td5JH4)
4. Your log's error : N/A 

(optional) If this is to do with permissions, 
5. Your permissions file :

(optional) If this is to do with chat,
6. Your chatconfig.yml : [https://pastebin.com/1K6HsCHN](https://pastebin.com/1K6HsCHN)
7. Your channels.yml : [https://pastebin.com/t2Z17EgY](https://pastebin.com/t2Z17EgY)

Tl;dr The event is 100% canceled when it gets to WG, but somewhere along the line Towny un-cancels it and DiscordSRV receives it as un-canceled.

**Comment by Arkobat** [comment]
We are planning to open our Towny server on November the first.
Is there any chance this will have a fix before this date, or will we need to drop TownyChat?
(Not only WG deny chat, but generally the `AsyncChatEvent#setCanceled(true)`

**Comment by LlmDl** [comment]
I can probably whip up a build for you to test things out with where we aren't ignorecancelling. Want it?

**Comment by LlmDl** [comment]
Try this out: https://cdn.discordapp.com/attachments/388338495036456960/630861475925458944/TownyChat-0.57-SNAPSHOT.jar

Tell me what happens with the cancelling, if there's any other side effects that appear in regular TC usage.

**Comment by Arkobat** [comment]
I’ll try it, and let you know How it tuens out!
Thanks :)

**Comment by Arkobat** [comment]
```
[18:18:12] [Server thread/ERROR]: Could not load 'plugins/TownyChat-0.57-SNAPSHOT.jar' in folder 'plugins'
org.bukkit.plugin.InvalidPluginException: Unsupported API version 1.14
	at org.bukkit.craftbukkit.v1_13_R2.util.CraftMagicNumbers.checkSupported(CraftMagicNumbers.java:226) ~[patched_1.13.2.jar:git-Paper-565]
	at org.bukkit.plugin.java.JavaPluginLoader.loadPlugin(JavaPluginLoader.java:126) ~[patched_1.13.2.jar:git-Paper-565]
	at org.bukkit.plugin.SimplePluginManager.loadPlugin(SimplePluginManager.java:330) ~[patched_1.13.2.jar:git-Paper-565]
	at org.bukkit.plugin.SimplePluginManager.loadPlugins(SimplePluginManager.java:252) ~[patched_1.13.2.jar:git-Paper-565]
	at org.bukkit.craftbukkit.v1_13_R2.CraftServer.loadPlugins(CraftServer.java:331) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.DedicatedServer.init(DedicatedServer.java:235) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:786) ~[patched_1.13.2.jar:git-Paper-565]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_77]
```
Towny version: 0.93.0.0

**Comment by LlmDl** [comment]
You should be on 0.94.0.0 with 1.13.2, but for testing's sake just edit the plugin.yml in the townychat.jar and set api-version to 1.13

**Comment by LlmDl** [comment]
I just tested this myself using 0.94.0.16, that Townychat snapshot and WG 7.0.1 and it works and I can't see any other side effects. I'll be interested to see what you get with further testing.

**Comment by Arkobat** [comment]
Sorry for the late response, I've had some things to do the last couple of days.  

Whenever I try to use a TownyChat command (like /g), I get an error.
Still use the same versions as before, but I've changed api version as told :)

```[17:05:46] [Server thread/INFO]: Arkobat issued server command: /g
[17:05:46] [Server thread/ERROR]: null
org.bukkit.command.CommandException: Unhandled exception executing 'g' in com.palmergames.bukkit.TownyChat.Command.commandobjects.ChannelJoinAliasCommand(g)
	at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:159) ~[patched_1.13.2.jar:git-Paper-565]
	at org.bukkit.craftbukkit.v1_13_R2.CraftServer.dispatchCommand(CraftServer.java:734) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.PlayerConnection.handleCommand(PlayerConnection.java:1798) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.PlayerConnection.a(PlayerConnection.java:1606) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.PacketPlayInChat.a(PacketPlayInChat.java:45) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.PacketPlayInChat.a(PacketPlayInChat.java:5) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:15) ~[patched_1.13.2.jar:git-Paper-565]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_77]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_77]
	at net.minecraft.server.v1_13_R2.SystemUtils.a(SystemUtils.java:108) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.MinecraftServer.b(MinecraftServer.java:1011) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.DedicatedServer.b(DedicatedServer.java:439) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:938) ~[patched_1.13.2.jar:git-Paper-565]
	at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:836) ~[patched_1.13.2.jar:git-Paper-565]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_77]
Caused by: java.lang.NoClassDefFoundError: com/palmergames/bukkit/towny/TownyUniverse
	at com.palmergames.bukkit.TownyChat.Command.commandobjects.ChannelJoinAliasCommand.execute(ChannelJoinAliasCommand.java:73) ~[?:?]
	at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:149) ~[patched_1.13.2.jar:git-Paper-565]
	... 14 more
Caused by: java.lang.ClassNotFoundException: com.palmergames.bukkit.towny.TownyUniverse
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381) ~[?:1.8.0_77]
	at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:138) ~[patched_1.13.2.jar:git-Paper-565]
	at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:84) ~[patched_1.13.2.jar:git-Paper-565]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[?:1.8.0_77]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_77]
	at com.palmergames.bukkit.TownyChat.Command.commandobjects.ChannelJoinAliasCommand.execute(ChannelJoinAliasCommand.java:73) ~[?:?]
	at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:149) ~[patched_1.13.2.jar:git-Paper-565]
	... 14 more```

**Comment by LlmDl** [comment]
Sorry, the TownyChat 0.56 and newer needs to be used with 0.94.0.2+, when we changed the Towny API.

Until you can get onto 1.14 and use a newer Towny you won't be able to use the chat cancelling.

**Comment by Arkobat** [comment]
1.14 will not even be considered, as the performance simply is to bad.
Is there any alternatives to TownyChat then?

**Comment by LlmDl** [comment]
All kinds, deluxe chat seems to be popular.

**Comment by LlmDl** [comment]
Closing because this is completed.

**Comment by delbertina** [comment]
I'm confused, what is the final verdict on this issue I submitted? Support for canceled chat events won't be added?

**Comment by LlmDl** [comment]
It was added as of 0.57.

**Comment by delbertina** [comment]
Is there something that needs to be changed in the configs for this to take effect? I'm seeing no change in behavior when upgraded to TC 0.58 and latest WG 7.0.1-RC2.

**Comment by LlmDl** [comment]
According to the discussion in this https://github.com/TownyAdvanced/Towny/issues/3423#issuecomment-544251633 ticket there might be some more work required. I don't know why it would work for me but not others unless there's other chat plugins influencing the priority that the town and nation channels run on.

**Comment by delbertina** [comment]
As stated in my initial post on this issue, I'm using no other chat plugin but TC. Without TC I get the expected behavior so does not seem to be an interference in my reproduction of the issue. 

This seems to be a similar issue that I remember some Factions plugin where every chat event was being canceled by the plugin. TC seems to be doing the opposite where every chat ends up un-canceled. I cant remember exactly if it was FactionsUUID or not, but maybe that helps you get closer to the root cause of this?

**Comment by LlmDl** [comment]
Yeah when I tested it, I was using only TC, and it cancelled chat just fine for me. 

I'll retest it, and see if it needs Arkobat's bits of code added to it to work properly.

**Comment by LlmDl** [comment]
@delbertina https://github.com/TownyAdvanced/TownyChat/releases/tag/0.60 That should cover it.

**Comment by delbertina** [comment]
Sorry for the long delayed response. Vacation + planning + catching up at work after kinda put a lot of stuff on the back burner.

Initial testing on my end seems to also reflect the 0.60 jar fixing the issue! I'll continue to monitor things to make sure there arnt any side effects. Thanks a million for helping me! I very much so appreciate it! : D <3
